osutil/sys/syscall.go

```diff
@@ -107,3 +109,61 @@ func FcntlGetFl(fd int) (int, error) {
 	}
 	return int(flags), nil
 }
+
+// UnrecoverableError is an error that flags that things have Gone Wrong, the
+// runtime is in a bad state, and you should really quit. The intention is that
+// if you're trying to recover from a panic and find that the value of the panic
+// is an UnrecoverableError, you should just exit ASAP.
+type UnrecoverableError struct {
+	Call string
+	Err  error
+}
+
+func (e UnrecoverableError) Error() string {
+	return fmt.Sprintf("%s: %v", e.Call, e.Err)
+}
+
+// RunAsUidGid starts a goroutine, pins it to the OS thread, sets euid and egid,
+// and runs the function; after the function returns, it restores euid and egid.
+//
+// If restoring the original euid and egid fails this function will panic with
+// an UnrecoverableError, and you should _not_ try to recover from it: the
+// runtime itself is going to be in trouble.
+func RunAsUidGid(uid UserID, gid GroupID, f func() error) error {
+	ch := make(chan error, 1)
+	go func() {
+		// from the docs:
+		//   until the goroutine exits or calls UnlockOSThread, it will
+		//   always execute in this thread, and no other goroutine can.
+		// that last bit means it's safe to setuid/setgid in here, as no
+		// other code will run.
+		runtime.LockOSThread()
+
+		ruid := Getuid()
+		rgid := Getgid()
+
+		if _, _, errno := syscall.RawSyscall(_SYS_SETREGID, FlagID, uintptr(gid), 0); errno == 0 {
+			if _, _, errno := syscall.RawSyscall(_SYS_SETREUID, FlagID, uintptr(uid), 0); errno == 0 {
+				ch <- f()
+				// try to restore euid
+				if _, _, errno := syscall.RawSyscall(_SYS_SETREUID, FlagID, uintptr(ruid), 0); errno != 0 {
+					// ¯\_(ツ)_/¯
+					panic(UnrecoverableError{Call: "setreuid", Err: errno})
```

Comment:
I'm actually not sure if we need to panic here, it seems risky in library code like this. I suspect we could just return the error here via the channel (and transmit an error from the function too, something like "fmt.Error("cannot setreuid: %v after function error %v", err, funcErr)") and remove UnrecoverableError type ?- or am I missing something?

Reply:
I think we want to prevent any further execution as soon as possible. There's many cases of unchecked errors from uid-swapping routines causing remote privilege escalations.

Reply:
The issue also being that no-one is listening for the error on the channel should restoring the gid/uid fail (as we already signalled the channel from the function completing)